internal/db_storage/db_service.go

```diff
@@ -0,0 +1,330 @@
+package dbstorage
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"log"
+	"time"
+
+	"github.com/developerc/project_gophermart_2/internal/general"
+	_ "github.com/jackc/pgx/v5/stdlib"
+)
+
+type ErrorLgnPsw struct {
+	s string
+}
+
+func (e *ErrorLgnPsw) Error() string {
+	return e.s
+}
+
+func (e *ErrorLgnPsw) AsLgnPswWrong(err error) bool {
+	return errors.As(err, &e)
+}
+
+func CreateTables(db *sql.DB) error {
+	const duration uint = 20
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(duration)*time.Second)
+	defer cancel()
+	const crusr string = "CREATE TABLE IF NOT EXISTS usr_table( uuid serial primary key, " +
+		"usr TEXT CONSTRAINT must_be_different_usr UNIQUE, psw TEXT)"
+	_, err := db.ExecContext(ctx, crusr)
+	if err != nil {
+		return err
+	}
+
+	const crord string = "CREATE TABLE IF NOT EXISTS orders_table( uuid serial primary key, " +
+		"usr TEXT, order_numb TEXT CONSTRAINT must_be_different_order UNIQUE, status TEXT NOT NULL DEFAULT 'NEW', accrual REAL NOT NULL DEFAULT 0.0, withdraw REAL NOT NULL DEFAULT 0.0, date_time TIMESTAMP NOT NULL DEFAULT NOW())"
+	_, err = db.ExecContext(ctx, crord)
+	if err != nil {
+		return err
+	}
+
+	const pgcrpt string = "CREATE EXTENSION pgcrypto"
+	_, err = db.ExecContext(ctx, pgcrpt)
+	if err != nil {
+		log.Println(err)
+	}
+
+	return nil
+}
+
+func InsertUser(db *sql.DB, usr, psw string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
+	defer cancel()
+	_, err := db.ExecContext(ctx, "INSERT INTO usr_table (usr, psw) values ($1, crypt($2, gen_salt('md5')))", usr, psw)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func CheckLgnPsw(db *sql.DB, usr, psw string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 600*time.Second)
+	defer cancel()
+	rows, err := db.QueryContext(ctx, "SELECT (psw = crypt($2, psw)) AS password_match FROM usr_table WHERE usr = $1 ", usr, psw)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	cntrRows := 0
+	for rows.Next() {
+		cntrRows++
+		var passwordMatch bool
+		err = rows.Scan(&passwordMatch)
+		if err != nil {
+			return err
+		}
+		if !passwordMatch {
+			return &ErrorLgnPsw{"login or password is not valid"}
+		}
+	}
+	err = rows.Err()
+	if err != nil {
+		return err
+	}
+	if cntrRows == 0 {
+		return &ErrorLgnPsw{"login or password is not valid"}
+	}
+
+	return nil
+}
+
+func UploadOrder(db *sql.DB, usr, orderNum string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	tx, err := db.Begin()
+	if err != nil {
+		return err
+	}
+	rows, err := db.QueryContext(ctx, "SELECT usr FROM orders_table WHERE order_numb = $1 ", orderNum)
```

Comment:
Транзакция в данной функции не захватывает SELECT и, соответственно, не имеет смысла, так как фактически прикрывает единственную операцию INSERT. Если стоит цель увязать между собой SELECT и INSERT, то запрос на получение данных необходимо дополнить FOR UPDATE
```suggestion
	rows, err := db.QueryContext(ctx, "SELECT usr FROM orders_table WHERE order_numb = $1 FOR UPDATE", orderNum)
```

Reply:
Прочитал что просто SELECT внутри транзакции может столкнуться с race condition. Исправил.